Tests/DatadogTests/Datadog/Mocks/ServerMock.swift

```diff
@@ -32,29 +32,28 @@ private class ServerMockProtocol: URLProtocol {
     }
 
     override func startLoading() {
-        guard let serverMock = ServerMock.activeInstance else {
-            preconditionFailure("Request was started while no `ServerMock` instance is active.")
```

Comment:
this part is not clear: what does it mean for us if `activeInstance` is `nil`? it doesn't fail anymore but is it a success?

Reply:
There is only one global session instance: `URLSession.serverMockURLSession`, configured with the `ServerMockProtocol`. When `ServerMock` instance is created it sets itself to static `ServerMock.activeInstance` weak property. The protocol talks to the `server` through this `activeInstance`.

If the `activeInstance` is `nil` it basically means the `ServerMock` was destroyed, so the test completed and there is no other running test which created the `server`.

Now, I made this change because some tests were failing on:
```swift
// ServerMock

deinit {
    precondition(Thread.isMainThread, "`ServerMock` should be deinitialized on the main thread.")
}
```

As this `func startLoading()` is the only place where we capture strong reference to the `ServerMock` on a background thread, this is probably the suspicious place causing the flakiness. My bet is that we end up some test from one of the client completion calls below, e.g. `client?.urlProtocol(self, didLoad: data)`, which completes the test, releases the reference to `ServerMock` on the test's thread, but the captured reference here in `startLoading()` still remains, causing the deinit off the main thread.

---

Bottom line: I don't like the current setup of `ServerMock` and the shared `URLSession.serverMockSession` approach. What we should probably do instead is creating one `URLSession` for each `ServerMock`. This could be done in `RUMM-417`.

Reply:
in short `activeInstance == nil` is not an expected case and we are not sure how we end up there.?
yes let's see in 417 👌

Reply:
Yes, it's not an expected state 👌.